pandas/core/internals/managers.py

```diff
@@ -252,6 +252,9 @@ def add_references(self, mgr: BaseBlockManager) -> None:
         Adds the references from one manager to another. We assume that both
         managers have the same block structure.
         """
+        if len(self.blocks) != len(mgr.blocks):
```

Comment:
I also don't understand how the case in transpose can ever run into this? Because we should only get here in case of `self._can_fast_transpose`, and then by definition there should only be 1 block?

Reply:
```
df = DataFrame({"a": [pd.Timedelta("1 days"), pd.NaT, pd.Timedelta("2 days")]}, dtype="object")

df.T
```

the constructor infers this to date time after transposing the array. We could consider this a bug though...

Reply:
Yeah we are calling maybe_infer_datetimelike under the hood which causes this.

Reply:
Should we deprecate this first?

Reply:
> I seem to recall that we have had discussions about this corner case before (IMO we should indeed change that)

And note that this is not only about transpose, it's about constructors in general. So the question is also if we want to change this:

```
In [10]: df = DataFrame({"a": np.array([pd.Timedelta("1 days"), pd.NaT, pd.Timedelta("2 days")], dtype=object)})

In [11]: df
Out[11]: 
       a
0 1 days
1    NaT
2 2 days

In [12]: df.dtypes
Out[12]: 
a    timedelta64[ns]
dtype: object
```

(but that's for another issue)

Although of course we could fix this locally in transpose to specifically fix transpose to preserve the dtype.

Reply:
I think the constructor would have to be deprecated. Transpose is more of a bug fix imo because that we are using the inference rules of the constructor under the hood is more an implementation detail.

Reply:
One related issue: https://github.com/pandas-dev/pandas/issues/39117

It's also not clear we necessarily want to deprecate this for the constructors. But for transpose() I would certainly consider it a bug / something to change.

Reply:
Yeah constructors makes a bit of sense.

I'll try what's failing locally when changing this.

Reply:
We have tests covering exactly this behavior :(

```
FAILED pandas/tests/frame/methods/test_transpose.py::TestTranspose::test_transpose_tzaware_2col_mixed_tz - AssertionError: Attributes of DataFrame.iloc[:, 0] (column name="A") are different
FAILED pandas/tests/frame/methods/test_transpose.py::TestTranspose::test_transpose_object_to_tzaware_mixed_tz - assert False
```

#26285

So lets merge this without tackling this case